### PR TITLE
Medusa compaction

### PR DESCRIPTION
**Pull request by jlhughes**
Compaction process:

1. block
2. roll ledgers
3. update dagger keys - optional
4. unblock
5. write all daos to nodes in MedusaEntryDAO order

See CompactionDAO

### CODE REVIEW

File: src/foam/nanos/medusa/CompactionDAO.js [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+      args: 'X x',
+      javaCode: `
+      final Logger logger = Loggers.logger(x, this);
+      // logger.info("rollNodes");
+
+      final ClusterConfigSupport support = (ClusterConfigSupport) x.get("clusterConfigSupport");
+      final ClusterConfig myConfig = support.getConfig(x, support.getConfigId());
+
+      AssemblyLine line = new AsyncAssemblyLine(x, null, support.getThreadPoolName());
+      final Set replies = new HashSet();
```
Comment (xuerongNanopay) [inline src/foam/nanos/medusa/CompactionDAO.js]: Should it use thread-safe set? Collections.synchronizedSet(new HashSet<>());

File: src/foam/nanos/medusa/CompactionDAO.js [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+      long waited = 0L;
+      long sleep = 1000L;
+      while ( waited < getMaxWait() ) {
+        try {
+          Thread.currentThread().sleep(sleep);
+          waited += sleep;
+        } catch (InterruptedException e) {
+          break;
+        }
+        if ( replies.size() >= nodes.size() -1 ) {
```
Comment (xuerongNanopay) [inline src/foam/nanos/medusa/CompactionDAO.js]: why here is `>= nodes.size() -1 ` not `>= nodes.size()`

Comment (jlhughes) [reply src/foam/nanos/medusa/CompactionDAO.js]: corrected.

File: src/foam/nanos/medusa/CompactionDAO.js (hunk 1 above)
Comment (jlhughes) [reply src/foam/nanos/medusa/CompactionDAO.js]: Refactored to use map.

**Pull request merged**